File: python/Composite/menu.py
from abc import ABC, abstractmethod
from typing import Deque, List
from collections import Iterable, deque


class MenuComponent(ABC):
    def print(self):
        pass


class MenuItem(MenuComponent):
    def __init__(
        self, name: str, description: str, vegetarian: bool, price: float
    ) -> None:
        self.name: str = name
        self.description: str = description
        self.vegetarian: bool = vegetarian
        self.price: float = price

    def print(self):
        print(f"{self.name}, {self.price} -- {self.description}")


class Menu(MenuComponent):
    def __init__(self, name, description) -> None:
        self.menu_components: List[MenuComponent] = []
        self.name: str = name
        self.description: str = description

    def add(self, menu_com: MenuComponent):
        self.menu_components.append(menu_com)

    def remove(self, menu_com: MenuComponent):
        self.menu_components.remove(menu_com)

    def print(self):
        print()
        print(f"{self.name}, {self.description}")
        print("--------------------------------")
        for menu_com in self.menu_components:
            menu_com.print()


class PancakeHouseMenu(Menu):
    def __init__(self) -> None:
        super().__init__("PANCAKE HOUSE MENU", "Breakfast")
        self.add(
            MenuItem(
                "K&P Pancake Breakfast",
                "Pancakes with scrambled eggs, and toast",
                True,
                2.99,
            )
        )
        self.add(
            MenuItem(
                "Regular Pancake Breakfast",
                "Pancakes with fried eggs,sausage",
                False,
                2.99,
            )
        )
        self.add(
            MenuItem(
                "Blueberry Pancake", "Pancakes made with fresh blueberries", True, 3.49
            )
        )
        self.add(
            MenuItem(
                "Waffles",
                "Waffles, with your choice of blueberries or strawberries",
                True,
                3.59,
            )
        )


class DinerMenu(Menu):
    def __init__(self) -> None:
        super().__init__("DINER MENU", "Lunch")
        self.add(
            MenuItem(
                "Vegetarian BLT",
                "(Fakin')Bacon with lettuce & tomato on whole wheat",
                True,
                2.99,
            )
        )
        self.add(
            MenuItem("BLT", "Bacon with lettuce & tomato on whole wheat", False, 2.99)
        )
        self.add(
            MenuItem(
                "Soup of the day",
                "Soup of the day, with a side of potato salad",
                False,
                3.29,
            )
        )
        self.add(
            MenuItem(
                "Hotdog",
                "A hot dog, with saurkraut, relish, onions, topped with cheese",
                False,
                3.05,
            )
        )


class CafeMenu(Menu):
    def __init__(self) -> None:
        super().__init__("CAFE MENU", "Dinner")
        self.add(
            MenuItem(
                "Veggie Burger and Air Fries",
                "Veggie burger on a whole wheat bun, lettuce, tomato, and fries",
                True,
                3.99,
            )
        )
        self.add(
            MenuItem(
                "Soup of the day",
                "A cup of the soup of the day, with a side salad",
                False,
                3.69,
            )
        )
        self.add(
            MenuItem(
                "Burrito",
                "A large burrito, with whole pinto beans, salsa, guacamole",
                True,
                4.29,
            )
        )


class DessertMenu(Menu):
    def __init__(self) -> None:
        super().__init__("DESSERT MENU", "Dessert of course")
        self.add(
            MenuItem(
                "Apple Pie",
                "Apple pie with a flakey crust, topped with vanilla icecream",
                True,
                1.59,
            )
        )
        self.add(
            MenuItem(
                "Cheesecake",
                "Creamy New York cheesecake, with a chocolate graham crust",
                True,
                1.99,
            )
        )
        self.add(
            MenuItem("Sorbet", "A scoop of raspberry and a scoop of lime", True, 1.89)
        )

File: c/Composite/menu.c
#include "menu.h"

#define MAX_MENU_ITEM_NUM 20
MenuItem menu_items[MAX_MENU_ITEM_NUM];
int menu_item_len = 0;

static void _add(struct Menu *self, MenuComponent *menu_com);
static void _print(struct Menu *self);
static MenuComponent *_create_menu_item(str name, str description, bool vegetarian, double price);

void MenuNew(Menu *self, str name, str description)
{
    self->add = _add;
    self->print = _print;
    self->name = name;
    self->description = description;
    self->menu_com_length = 0;
}

static void _add(struct Menu *self, MenuComponent *menu_com)
{
    self->menu_coms[self->menu_com_length] = menu_com;
    self->menu_com_length++;
}

static void _print(struct Menu *self)
{
    printf("\n%s, %s\n", self->name, self->description);
    printf("--------------------------------\n");
    //printf("self->menu_com_length = %s\n", self->menu_com_length);

    for (int i = 0; i < self->menu_com_length; i++)
    {
        self->menu_coms[i]->print(self->menu_coms[i]);
    }
}

static MenuComponent *_create_menu_item(str name, str description, bool vegetarian, double price)
{
    MenuItemNew(&menu_items[menu_item_len], name, description, vegetarian, price);
    menu_item_len++;
    return (MenuComponent *)&menu_items[menu_item_len - 1];
}

void PancakeHouseMenuNew(Menu *self)
{
    MenuNew(self, "PANCAKE HOUSE MENU", "Breakfast");
    MenuComponent *menu_com = _create_menu_item("K&P Pancake Breakfast", "Pancakes with scrambled eggs,toast", true, 2.99);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Regular Pancake Breakfast", "Pancakes with fried eggs,sausage", false, 2.99);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Blueberry Pancake", "Pancakes made with fresh blueberries", true, 3.49);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Waffles", "Waffles, with your choice of blueberries or strawberries", true, 3.59);
    self->add(self, menu_com);
}

void DinerMenuNew(Menu *self)
{
    MenuNew(self, "DINER MENU", "Lunch");
    MenuComponent *menu_com = _create_menu_item("Vegetarian BLT",
                                                "(Fakin')Bacon with lettuce & tomato on whole wheat",
                                                true,
                                                2.99);
    self->add(self, menu_com);
    menu_com = _create_menu_item("BLT", "Bacon with lettuce & tomato on whole wheat", false, 2.99);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Soup of the day",
                                 "Soup of the day, with a side of potato salad",
                                 false,
                                 3.29);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Hotdog",
                                 "A hot dog, with saurkraut, relish, onions, topped with cheese",
                                 false,
                                 3.05);
    self->add(self, menu_com);
}

void CafeMenuNew(Menu *self)
{
    MenuNew(self, "CAFE MENU", "Dinner");
    MenuComponent *menu_com = _create_menu_item("Veggie Burger and Air Fries",
                                                "Veggie burger on a whole wheat bun, lettuce, tomato, and fries",
                                                true,
                                                3.99);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Soup of the day",
                                 "A cup of the soup of the day, with a side salad",
                                 false,
                                 3.69);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Burrito",
                                 "A large burrito, with whole pinto beans, salsa, guacamole",
                                 true,
                                 4.29);
    self->add(self, menu_com);
}

void DessertMenuNew(Menu *self)
{
    MenuNew(self, "DESSERT MENU", "Dessert of course");
    MenuComponent *menu_com = _create_menu_item("Apple Pie",
                                                "Apple pie with a flakey crust, topped with vanilla icecream",
                                                true,
                                                1.59);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Cheesecake",
                                 "Creamy New York cheesecake, with a chocolate graham crust",
                                 true,
                                 1.99);
    self->add(self, menu_com);
    menu_com = _create_menu_item("Sorbet", "A scoop of raspberry and a scoop of lime", true, 1.89);
    self->add(self, menu_com);
}